### PR TITLE
Add pybind11

### PR DESCRIPTION
**Pull request by mwoehlke-kitware**
EWISOTT

Jamie/Robin for feature review, Jeremy/David for platform review.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/4924)

### CODE REVIEW

File: cmake/options.cmake [hunk 1]
```diff
@@ -233,6 +233,11 @@ macro(drake_setup_options)
     "Google command-line flags processing library")
 
   drake_system_dependency(
```
Comment (mwoehlke-kitware) [inline cmake/options.cmake]: FYI: This implies that pybind11 is required (unless Python is disabled), but you can choose to use the system version. I'm happy to adjust if folks aren't happy with that arrangement.

**Comment by david-german-tri** [comment]
:lgtm:

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-:-KbRhRKgmuiu7mulL3Zf:bnfp4nl)*

**Comment by jwnimmer-tri** [comment]
:lgtm:

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[CMakeLists.txt, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-KbRgvlLJHMVD6rS7PJg:-KbRgvlLJHMVD6rS7PJh:bvaukze) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5ac18e00f080a367ae18b312e3b67aaf092be3/CMakeLists.txt#L131)):*
> ```CMake
> 
> # pybind11
> drake_add_external(pybind11 CMAKE
> ```

Should this say `PYTHON`?  Or does the "Uses Python" not count for the Python C APIs?

---

*[cmake/options.cmake, line 235 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-KbReuQklY7WcMJQwAMy-r1-235:-KbRg0mUUb9UgZ-Gp35_:b-81nlyc) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5ac18e00f080a367ae18b312e3b67aaf092be3/cmake/options.cmake#L235)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

FYI: This implies that pybind11 is required (unless Python is disabled), but you can choose to use the system version. I'm happy to adjust if folks aren't happy with that arrangement.
</blockquote></details>

SGTM.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-:-KbRgK1NIAB8RJMgf6VQ:bnfp4nl)*

**Comment by david-german-tri** [comment]
Please add support to the Bazel build as well, but a separate PR after everything's up and running in CMake is fine.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-:-KbRhUj-XIdMPLdp2COg:bd7ng96)*

**Comment by jamiesnape** [comment]
:lgtm:

(I have Bazel build files already so I can PR that momentarily)

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-:-KbRiJPnO_irNfyuX6jc:b-6x0ebl)*

**Comment by jamiesnape** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[CMakeLists.txt, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-KbRgvlLJHMVD6rS7PJg:-KbRj2gPER8TpSeXkL-h:bgxyxvd) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5ac18e00f080a367ae18b312e3b67aaf092be3/CMakeLists.txt#L131)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Should this say `PYTHON`?  Or does the "Uses Python" not count for the Python C APIs?
</blockquote></details>

If it looks for either the python executable or its libraries during the configure, it quite possibly should.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924)*

**Comment by rdeits** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-:-KbRjU5n2g5rYlWQ3jb9:bnfp4nl)*

**Comment by mwoehlke-kitware** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[CMakeLists.txt, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-KbRgvlLJHMVD6rS7PJg:-KbRjxKMznXG5G-CbGTQ:b5wkrmg) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5ac18e00f080a367ae18b312e3b67aaf092be3/CMakeLists.txt#L131)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

If it looks for either the python executable or its libraries during the configure, it quite possibly should.
</blockquote></details>

That is a good question, and @jamiesnape just (indirectly) explained why the answer is "no" :smile:. (Well... it _could_, but it would be superfluous.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924)*

**Comment by mwoehlke-kitware** [comment]
I don't think we can know if `PYBIND11_PYTHON_VERSION` is working right until someone tries to actually *use* this, so if there are no objections by tomorrow¹, I'll assume that means no one has changed their mind about their LGTM :smile:.

(¹ ...because it's late enough I am probably leaving before CI finishes.)

---

Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-:-KbRlK9LCULCIlf3GP06:bf71z6y)*

**Comment by rdeits** [comment]
I can confirm that the `PYBIND11_PYTHON_VERSION` setting does the right thing. I just used it a couple minutes ago to switch my local IRIS build from python3.5 to python2.7, and I'm using it on travis. Note that you *don't* need to pass that variable to the pybind11 project, just to the project that uses it (i.e. Drake)

---

Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-:-KbRmBZIVHU-Lmn3RSSd:bkobhax)*

File: cmake/options.cmake (hunk 1 above)
Comment (jamiesnape) [inline cmake/options.cmake]: I think I would prefer us to always use the Drake version, but I will defer to whatever everyone else decides.

**Comment by rdeits** [comment]
:lgtm:

---

Reviewed 3 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-:-KbRmXabrk88CqXn-Ani:bnfp4nl)*

**Comment by jwnimmer-tri** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924)*

**Comment by jamiesnape** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924)*

**Comment by mwoehlke-kitware** [comment]
> Note that you *don't* need to pass [`PYBIND11_PYTHON_VERSION`] to the pybind11 project, just to the project that uses it (i.e. Drake)

Huh... I guess pybind11 must include itself in its CMake. I was "confused" because that variable shows up in the configure for pybind11 itself...

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[cmake/options.cmake, line 235 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-KbReuQklY7WcMJQwAMy-r1-235:-KbRrzIUIMMI4lkYHhGJ:b-gr9tng) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5ac18e00f080a367ae18b312e3b67aaf092be3/cmake/options.cmake#L235)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

I think I would prefer us to always use the Drake version, but I will defer to whatever everyone else decides.
</blockquote></details>

We don't really encourage people to use non-Drake-provided versions of "system" dependencies (except VTK), but we also historically have not really supported the concept of non-optional dependencies (although I added those when "flattening" director). If anyone feels strongly, I could change this to a `drake_dependent_external`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4924#-:-KbRrhQe0bwKBUDer-5O:benk75p)*

**Comment by mwoehlke-kitware** [comment]
(Merged as-is; we can make further changes in a separate PR if desired.)

**Pull request merged**